pkgs/build-support/install-shell-files/default.nix

```diff
@@ -0,0 +1,31 @@
+{ lib, makeSetupHook }:
+
+(makeSetupHook { name = "install-shell-files"; } ../setup-hooks/install-shell-files.sh) // {
+  # This package should be callable as well in order to pass shell completions to the shell hook.
+  # See the header comment in ../setup-hooks/install-shell-files.sh for examples.
+  __functor = _: arg:
+    if builtins.isList arg then
+      # This is a convenience for escaping each manpages entry.
+      map lib.escapeShellArg arg
+    else if builtins.isString arg then
+      # Assume this is a manpages value without the list.
+      lib.escapeShellArg arg
+    else # must be an attrset
+      # Value looks like:
+      #   {
+      #     bash = [ "share/completions/foobar.bash" ];
+      #     fish = [ "share/completions/foobar.fish" ];
+      #   }
+      #
+      # For convenience, let's also accept
+      #  {
+      #    bash = "share/completions/foobar.bash";
+      #    fish = "share/completions/foobar.fish";
+      #  }
+      let
+        escape = arg: lib.replaceStrings ["\\" "\n"] ["\\\\" "\\n"] arg;
+      in
+      lib.concatStrings (lib.attrsets.mapAttrsToList (name: value:
+        lib.strings.concatMapStrings (path: "${escape name}=${escape path}\n") (lib.lists.toList value)
+      ) arg);
```

Comment:
I'm thinking I should tweak this so it just tacks on a `__tostring` attribute that does the actual conversion. That way you can query the structured shell completions on the Nix side and it only produces the Bash version when serialized into the environment.